### ISSUE TITLE
cat /dev/urandom | sleep 1    hangs

### ISSUE THREAD

**Issue by jyn514** [oilshell/oil#295]
Example: `/bin/cat /dev/urandom | head > /dev/null`. `bash` and `dash` will kill /bin/cat after less than a second, `osh` runs it indefinitely, you have to interrupt the process. We need to handle SIGCHILD. This is related to https://github.com/oilshell/oil/issues/176.

strace of `dash`:

```
sudo strace -f -p 853 -e 'trace=open,read,write'
strace: Process 853 attached
read(0, "/bin/cat /dev/urandom | head > /"..., 8192) = 41
strace: Process 1330 attached
strace: Process 1331 attached
pid  1330] read(3,  <unfinished ...>
[pid  1331] read(0,  <unfinished ...>
[pid  1330] <... read resumed> "c\203<Oeq\n\273\305\23H\317,\7\211\213-R\267\210\347\227\343\226y\301\0039\326\3655\234"..., 131072) = 131072
[pid  1330] write(1, "c\203<Oeq\n\273\305\23H\317,\7\211\213-R\267\210\347\227\343\226y\301\0039\326\3655\234"..., 131072 <unfinished ...>
[pid  1331] <... read resumed> "c\203<Oeq\n\273\305\23H\317,\7\211\213-R\267\210\347\227\343\226y\301\0039\326\3655\234"..., 8192) = 8192
[pid  1330] <... write resumed> )       = 73728
[pid  1331] write(1, "c\203<Oeq\n\273\305\23H\317,\7\211\213-R\267\210\347\227\343\226y\301\0039\326\3655\234"..., 2656 <unfinished ...>
[pid  1330] --- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=1330, si_uid=1000} ---
[pid  1331] <... write resumed> )       = 2656
[pid  1330] +++ killed by SIGPIPE +++
[pid   853] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=1330, si_uid=1000, si_status=SIGPIPE, si_utime=0, si_stime=1} ---
[pid  1331] +++ exited with 0 +++
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=1331, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
```

strace of osh:

```
[pid  2990] read(11, "\246\235\361\334\5\363\235\220\365z\346\204\223}<9\365\3552\257vX\320\250<\376\354`\27\27\201\365"..., 131072) = 131072
[pid  2990] write(1, "\246\235\361\334\5\363\235\220\365z\346\204\223}<9\365\3552\257vX\320\250<\376\354`\27\27\201\365"..., 131072 <unfinished ...>
[pid  2991] read(0, "\246\235\361\334\5\363\235\220\365z\346\204\223}<9\365\3552\257vX\320\250<\376\354`\27\27\201\365"..., 8192) = 8192
[pid  2991] close(0)                    = 0
[pid  2991] write(1, "\246\235\361\334\5\363\235\220\365z\346\204\223}<9\365\3552\257vX\320\250<\376\354`\27\27\201\365"..., 2737) = 2737
[pid  2991] close(1)                    = 0
[pid  2991] close(2)                    = 0
[pid  2991] +++ exited with 0 +++
[pid  2772] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=2991, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
[pid  2772] close(14)                   = 0
[pid  2772] close(12)
... hangs indefinitely ...
```

**Comment by jyn514** [comment]
So I actually had this wrong - SIGPIPE is handled by the kernel, the reason it gets sent is when the next process closes the file descriptor. Here's a better strace which includes `close`:

```
[pid  1991] write(1, "};m\4w\262\255\"!x\275\313\332\32\317\217\\\267.\270\34\242\216H:\177\360\207\214~\274{"..., 131072 <unfinished ...>
[pid  1992] close(3)                    = 0
[pid  1992] read(0, "};m\4w\262\255\"!x\275\313\332\32\317\217\\\267.\270\34\242\216H:\177\360\207\214~\274{"..., 8192) = 8192
[pid  1992] write(1, "};m\4w\262\255\"!x\275\313\332\32\317\217\\\267.\270\34\242\216H:\177\360\207\214~\274{"..., 4096) = 4096
[pid  1992] close(0)                    = 0
[pid  1991] <... write resumed> )       = 73728
[pid  1992] write(1, "\f{\226\211\212\37\307\23\316p\0310\311\207\311\"\202@\t\342v\355\350\261u\22\310\343\201\24329"..., 232) = 232
[pid  1992] close(1)                    = 0
[pid  1992] close(2)                    = 0
[pid  1992] +++ exited with 0 +++
[pid  1991] --- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=1991, si_uid=1000} ---
[pid  1763] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=1992, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
[pid  1991] +++ killed by SIGPIPE +++
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=1991, si_uid=1000, si_status=SIGPIPE, si_utime=0, si_stime=1} ---
```

Note how `cat` gets SIGPIPE right after `head` closes stdin. Not sure why this doesn't happen in osh, give me a minute.

**Comment by jyn514** [comment]
`man pipe` says

>If all file descriptors referring to the write end of a pipe have been closed, then
       an  attempt  to read(2) from the pipe will see end-of-file (read(2) will return 0).
       If all file descriptors referring to the read end of a pipe have been closed,  then
       a write(2) will cause a SIGPIPE signal to be generated for the calling process.  If
       the calling process is ignoring this signal, then write(2)  fails  with  the  error
       EPIPE.   An  application that uses pipe(2) and fork(2) should use suitable close(2)
       calls to close unnecessary duplicate file descriptors; this  ensures  that  end-of-
       file and SIGPIPE/EPIPE are delivered when appropriate.

Maybe the child forgets to close stdin?

**Comment by andychu** [comment]
Hm I reproduced the bug but I'm not sure what's going on yet!  Thanks for the report.

**Comment by jyn514** [comment]
More minimal test case: `cat /dev/null | sleep 1`

**Comment by andychu** [comment]
@jyn514 I just fixed this (see change description for the cause).  This was a good bug, keep them coming! :)

And signal handling is generally improving ... I'm getting away from the whack-a-mole strategy.

**Comment by andychu** [comment]
Fixed with 0.6.pre22: http://www.oilshell.org/release/0.6.pre22/changelog.html
